### PR TITLE
OSS Fuzz CMake changes

### PR DESCRIPTION
**Pull request by minafarid**
* Added `OSS_FUZZ` CMake option.
* Load the fuzzing library either from the library provided by OSS Fuzz or from `libFuzzer.a` that was manually built.
* This change allows running fuzz testing both locally using the CMake option `-DFUZZING=ON` and on OSS Fuzz using CMake option `-DOSS_FUZZ=ON`.

Update:
* No need to manually set `OSS_FUZZ` option, we only need to set `FUZZING=ON` both for local fuzzing and for OSS Fuzz. The CMake script detects the existence of a fuzzing library in `LIB_FUZZING_ENGINE` and uses it. Otherwise, it falls back to use `libFuzzer` that is built from sources.

### CODE REVIEW

File: Firestore/core/test/firebase/firestore/model/field_value_test.cc [hunk 1]
```diff
@@ -216,14 +216,14 @@ TEST(FieldValue, Copy) {
   clone = null_value;
   EXPECT_EQ(FieldValue::NullValue(), clone);
   EXPECT_EQ(FieldValue::NullValue(), null_value);
-  clone = clone;
+  clone = *&clone;
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/model/field_value_test.cc]: Can you split out this change into a separate PR (and add Zhi as a reviewer on that)? It would also be great to add a (single) brief comment with a link to Clang docs which explain the issue.

**Comment by minafarid** [comment]
@var-const 
I updated the original GitHub comment to remove the changes in `field_value_test.cc` and include the CMake options required for local fuzzing and OSS Fuzz.

**Comment by minafarid** [comment]
PR #1612 is meged. This PR is ready to be merged.

File: cmake/FindFuzzingLibrary.cmake [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# OSS_FUZZ provides its own fuzzing library in LIB_FUZZING_ENGINE environment
+# variable. For local fuzzing, search for the libFuzzer.a library that was
+# manually built.
+if(OSS_FUZZ)
+  set(FUZZING_LIBRARY_LOCATION $ENV{LIB_FUZZING_ENGINE})
```
Comment (minafarid) [inline cmake/FindFuzzingLibrary.cmake]: @wilhuff are we ok with reading from environment variables? OSS Fuzz builds the code inside a docker container with set environment variables.

File: cmake/external/libfuzzer.cmake [hunk 3]
```diff
@@ -27,6 +27,12 @@ if(TARGET libfuzzer)
   return()
 endif()
 
+# Mark libfuzzer target as done on Windows because fuzzing is not supported.
+if(WIN32)
+  add_custom_target(libfuzzer)
```
Comment (minafarid) [inline cmake/external/libfuzzer.cmake]: @wilhuff I just want to confirm that this line indeed marks `libfuzzer` target as done. We do not support fuzzing under Windows.

File: cmake/FindFuzzingLibrary.cmake (hunk 2 above)
Comment (wilhuff) [reply cmake/FindFuzzingLibrary.cmake]: Reading from environment variables like this is fine, however I have some feedback:
  * The naming of this package confuses and interchanges the words "fuzzer" and "fuzzing".
  * It also uses "library" as a suffix when CMake convention is to drop "lib" and "library" from find-package names. For example, `libjpeg` is available in `FindJPEG.cmake`.

It's unfortunate that OSS_FUZZ calls this `LIB_FUZZING_ENGINE` but they're apparently abstracting over what this thing is. We should plan for this to be concrete and just happen to accept their variable to configure this.

So, assuming the thing we build against is "fuzzer":

  * `$ENV{LIB_FUZZING_ENGINE}` should just be the first hint to `find_library`.
    * If it's set, we'll use it, otherwise we'll fall back on whatever we've built.
    * I don't think we need to condition this on `OSS_FUZZ` being set.
  * Since we're finding "libfuzzer" this should in `FindFuzzer.cmake`
  * The output variable of `find_library` should be `FUZZER_LIBRARY`
  * The first argument to `find_package_handle_standard_args` should also match, so it should be `FUZZER`
  * That will result in `FUZZER_FOUND` being set.

Also, to compile against the fuzzer, don't we need headers? You should also `find_path` on one of the headers (like we do for [nanopb](https://github.com/firebase/firebase-ios-sdk/blob/master/cmake/FindNanopb.cmake#L17)) and attach that to the `INTERFACE_INCLUDE_DIRECTORIES` property of this target.

File: cmake/external/libfuzzer.cmake (hunk 3 above)
Comment (wilhuff) [reply cmake/external/libfuzzer.cmake]: This is the correct way to do it.

You should also short-circuit this way if:
  * The fuzzing option is not set (there's no reason to build the library if we're not fuzzing)
  * OSS Fuzz is supplying its own pre-built fuzzer

This may require plumbing options from the outer build to the inner dependency fetching build.

File: cmake/fuzzing_options.cmake [hunk 4]
```diff
@@ -16,11 +16,18 @@
 
 include(compiler_id)
 
-option(FUZZING "Build for Fuzz Testing (local fuzzing and OSS Fuzz)" OFF)
+option(FUZZING "Build for local Fuzz Testing" OFF)
+option(OSS_FUZZ "Build for OSS Fuzz Environment" OFF)
```
Comment (wilhuff) [inline cmake/fuzzing_options.cmake]: Is this necessary? Could we detect that OSS Fuzz is in effect by the presence of their environment variable? (I'd suggest we still require FUZZING to be ON for either case.)

**Comment by minafarid** [comment]
I did most of the required changes. I changed the naming and detected `OSS_FUZZ` from the environment variable, when needed. Thanks for the `find_library` hint, it simplified the code. 

Two notes:
1. We do not need to include the headers of `libFuzzer` because we only link to it and provide an extern C function (like [serializer test here](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/core/src/firebase/firestore/fuzzing/fuzz_test_serializer.cc#L26)) and we use the main function of libFuzzer itself.
2. I could not pass the `FUZZING` option form the parent project to `cmake/external/libfuzzer.cmake`. How can we pass options to external projects? The negative effect of this is that we build libFuzzer even when `FUZZING` is not enabled. I added a TODO as well to do it next.

File: cmake/fuzzing_options.cmake (hunk 4 above)
Comment (minafarid) [reply cmake/fuzzing_options.cmake]: No not necessary, removed and detected from `ENV{LIB_FUZZING_ENGINE}` when needed.

File: cmake/external/libfuzzer.cmake (hunk 3 above)
Comment (minafarid) [reply cmake/external/libfuzzer.cmake]: How can we pass the `FUZZING` option when building external projects?

File: cmake/FindFuzzingLibrary.cmake (hunk 2 above)
Comment (minafarid) [reply cmake/FindFuzzingLibrary.cmake]: I did the renaming and `find_library` modifications. As I explained in my previous comment, we do not need headers from libFuzzer as we only link to it and provide the fuzzing function that is needed by libFuzzer's main function

File: cmake/external/libfuzzer.cmake (hunk 3 above)
Comment (wilhuff) [reply cmake/external/libfuzzer.cmake]: By adding to the `CMAKE_ARGS` you pass to `ExternalProject_Add`.

In this case, add `-DFUZZING=${FUZZING}` to the CMake invocation here: https://github.com/firebase/firebase-ios-sdk/blob/master/cmake/external_rules.cmake#L19

File: cmake/fuzzing_options.cmake [hunk 5]
```diff
@@ -16,8 +16,12 @@
 
 include(compiler_id)
 
-option(FUZZING "Build for local Fuzz Testing" OFF)
-option(OSS_FUZZ "Build for OSS Fuzz Environment" OFF)
+option(FUZZING "Build for Fuzz Testing (local fuzzing and OSS Fuzz)" OFF)
+
+# Assume OSS Fuzz if LIB_FUZZING_ENGINE environment variable is set.
+if(DEFINED ENV{LIB_FUZZING_ENGINE})
+  set(OSS_FUZZ ON)
```
Comment (wilhuff) [inline cmake/fuzzing_options.cmake]: `OSS_FUZZ` no longer seems to be in use aside from the check below. Maybe just remove it?

Comment (minafarid) [reply cmake/fuzzing_options.cmake]: Removed. I just thought it is more readable.

File: cmake/external/libfuzzer.cmake (hunk 3 above)
Comment (minafarid) [reply cmake/external/libfuzzer.cmake]: Done, thanks.

File: cmake/FindFuzzer.cmake [hunk 6]
```diff
@@ -12,26 +12,34 @@
...
-    )
-  endif()
-endif(LIBFUZZER_FOUND)
+if(FUZZER_FOUND)
+  add_library(Fuzzer STATIC IMPORTED)
+  set_target_properties(
+    Fuzzer PROPERTIES
+    IMPORTED_LOCATION ${FUZZER_LOCATION}
+  )
+else()
```
Comment (wilhuff) [inline cmake/FindFuzzer.cmake]: `find_package_handle_standard_args` already handles this for you because you're calling it via `find_package(Fuzzer REQUIRED)`. This `else()` can be removed.

Comment (minafarid) [reply cmake/FindFuzzer.cmake]: Good point, thanks. Removed.
